teps/0147-dynamic-matrix.md

```diff
@@ -0,0 +1,361 @@
+maintain a single pipeline in a catalog and allow the users to specify the list of explicit combinations for each
+instance dynamically through `pipelineRun` or a trigger template or a task result.
+
+The `matrix.strategy` is of type `string` where its value is a list of objects in key:value pairs. The controller reads
+a stringified JSON payload in a form of `{"include": a list of combinations}` and creates an equivalent number of
+instances based on the length of the specified list.
+
+When `matrix.strategy` is specified, no other `matrix` fields will be allowed i.e. `matrix.params` and `matrix.include`
+are not permitted with `matrix.strategy`. This restriction will help avoid any undesired conflicts after the `matrix`
+is resolved. 
+
+### `matrix.strategy` initialized by the `pipelineRun`
+
+The `Pipeline` addressing this use case will be defined as shown below:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  params:
+    - name: matrix-strategy
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: kaniko-build
+      taskRef:
+        name: kaniko
+      workspaces:
+        - name: source
+          workspace: shared-workspace
+      matrix:
+        strategy: $(params.matrix-strategy)
+```
+
+A `PipelineRun` can be created to execute the above `pipeline` to build 1 image:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: pipelineRun-
+spec:
+  pipelineRef:
+    name: pipeline-to-build-images-from-a-single-repo
+  params:
+    - name: matrix-strategy
+      value: "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"}]}"
+```
+
+`PipelineRun` can be created to execute the above `pipeline` to build 3 images:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: pipelineRun-
+spec:
+  pipelineRef:
+    name: pipeline-to-build-images-from-a-single-repo 
+  params:
+    - name: matrix-strategy
+      value: "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"},{\"IMAGE\":\"image-2\",\"DOCKERFILE\":\"path/to/Dockerfile2\"},{\"IMAGE\":\"image-3\",\"DOCKERFILE\":\"path/to/Dockerfile3\"}]}"
+```
+
+### `matrix.strategy` specified by a task result
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: pipeline-to-build-images-from-a-single-repo
+spec:
+  workspaces:
+    - name: shared-workspace
+  tasks:
+    - ...
+    - name: pre-requisite
+      taskSpec:
+        results:
+          - name: matrix-strategy
+        steps:
+          - name: read-strategy
+            image: alpine
+            script: |
+              # this strategy can be read from a JSON file in the repo
+              echo -n "{\"include\":[{\"IMAGE\":\"image-1\",\"DOCKERFILE\":\"path/to/Dockerfile1\"}]}" | tee $(results.matrix-strategy.path)
```

Comment:
Regular explicit combinations require a `name` field as well. Is the absence intentional?

Reply:
right it is intentional, optional `name` is supported with `include` section.`name` is for information purpose and has not real significance in how combinations are being generated. `name` is useful from the readability perspective when we have a few explicit combinations specified in the `pipeline` definition itself.

The goal of `strategy` is to be able to support dynamic combinations. 

I think, the JSON inherently does not support an optional specifications. The users have to specify empty string at the minimum. That's the reason it is not included in this specificaiton.

Reply:
got it, thanks (might be worth putting this in the TEP)

Reply:
Done, thanks!